File: python/src/day3/main.py
from src.common.file_utils import get_path, read_lines
from functools import reduce


def part_one(filename: str) -> int:
    lines = read_lines(get_path(__file__, filename))
    return get_trees(1, 3, lines)


def part_two(filename: str) -> int:
    lines = read_lines(get_path(__file__, filename))
    trees = [
        get_trees(1, 1, lines),
        get_trees(1, 3, lines),
        get_trees(1, 5, lines),
        get_trees(1, 7, lines),
        get_trees(2, 1, lines)
    ]
    return reduce((lambda x, y: x * y), trees)


def get_trees(down, right, lines):
    x, y, trees, width = 0, 0, 0, len(lines[0])
    while y < len(lines):
        trees += lines[y][x % width] == "#"
        x += right
        y += down
    return trees


if __name__ == '__main__':

    print("---Part One---")
    part_one_res = part_one("input.txt")
    print(part_one_res)

    print("---Part Two---")
    part_two_res = part_two("input.txt")
    print(part_two_res)

File: go/day3/main.go
package main

import (	
	input "aoc2020/inpututils"

	"fmt"
)

func main(){
	fmt.Println("--- Part One ---")
	fmt.Println(Part1("input.txt"))
	
	fmt.Println("--- Part Two ---")
	fmt.Println(Part2("input.txt"))
}

type treeCounter struct {
	lines []string
	h, w int
}

func newTreeCounter(lines []string) treeCounter {
	return treeCounter {
		lines: lines,
		w: len(lines[0]),
		h: len(lines),
	}
} 

func (tc treeCounter) count (dx int, dy int) int {
	x, y, trees := 0, 0, 0
	for y < tc.h {
		if string(tc.lines[y][x % tc.w]) == "#" {
			trees += 1
		}
		x += dx
		y += dy
	}
	return trees
}

func Part1(filename string) int {
	lines := input.ReadLines(filename)
	tc := newTreeCounter(lines)
	return tc.count(3, 1)
}

func Part2(filename string) int {
	lines := input.ReadLines(filename)
	tc := newTreeCounter(lines)
	return tc.count(1, 1) * tc.count(3, 1) * tc.count(5, 1) * tc.count(7, 1) * tc.count(1, 2)
}